### PR TITLE
Train data split & Early stopping

### PR DESCRIPTION
**Pull request by smguo**
This PR adds option in `train` to split train data into training and validation sets, and to early stop training if validation loss does not improve for user specified number of epochs. Default split is train/validation = 80/20.  

This PR is build on #20 and #21 and should be merged to master after both #20 and #21 have been merged

### CODE REVIEW

File: configs/config_example.yml [hunk 1]
```diff
@@ -0,0 +1,87 @@
+
+files:
+  raw_dirs: ['/CompMicro/projects/cardiomyocytes/200721_CM_Mock_SPS_Fluor/20200721_CM_Mock_SPS/dnm_input_tstack',
+             '/CompMicro/projects/cardiomyocytes/20200722CM_LowMOI_SPS_Fluor/20200722 CM_LowMOI_SPS/dnm_input_tstack']
+
+  supp_dirs: ['/CompMicro/projects/cardiomyocytes/200721_CM_Mock_SPS_Fluor/20200721_CM_Mock_SPS/dnm_supp_tstack',
+              '/CompMicro/projects/cardiomyocytes/20200722CM_LowMOI_SPS_Fluor/20200722 CM_LowMOI_SPS/dnm_supp_tstack']
+
```
Comment (mattersoflight) [inline configs/config_example.yml]: It seems that the config structure needs to be updated after PR #20 .

**Comment by mattersoflight** [comment]
Hi @smguo, how much effort would it be to use pytorch dataloader for batching and augmenting both training and validation sets (issue #15)? It will be good to train a model to make sure that code is in decent shape before merging.

**Comment by bryantChhun** [comment]
I have not much to say other than this PR seems well-isolated and ok to merge.  As long as it's up to date with the new config structure and master, i'm ok with it.

I will trust you've done enough testing to verify

**Review by bryantChhun — APPROVED** [review]
Just be sure it's synched with master.  Otherwise I think it's well isolated and shouldn't cause problems later

**Comment by smguo** [comment]
> Hi @smguo, how much effort would it be to use pytorch dataloader for batching and augmenting both training and validation sets (issue #15)? It will be good to train a model to make sure that code is in decent shape before merging.

@mattersoflight I did a test training run with small number of epochs and the new features work fine. Using dataloader to load batches from the hard drive can scale to large dataset but will require more changes because the patches need to be saved as individual images. I can implement dataloader class to sample patches from numpy arrays so we don't need to change the current patch file format

**Comment by smguo** [comment]
> Just be sure it's synched with master. Otherwise I think it's well isolated and shouldn't cause problems later

I have merged master into this PR and resolve conflicts. Will merge to master.
